sdk/clientcore/core/src/main/java/io/clientcore/core/util/union/Union.java

```diff
@@ -0,0 +1,284 @@
+
+    /**
+     * Creates a new instance of {@link Union} with the provided types.
+     *
+     * @param types The types of the union.
+     * @return A new instance of {@link Union}.
+     */
+    public static Union ofTypes(Type... types) {
+        return new Union(types);
+    }
+
+    /**
+     * Sets the value of the union.
+     *
+     * @param value The value of the union.
+     * @throws IllegalArgumentException If the value is not of one of the types in the union.
+     */
+    @SuppressWarnings("unchecked")
+    public void setValue(Object value) {
+        for (Type type : types) {
+            if (isInstanceOfType(value, type) || isPrimitiveTypeMatch(value, type)) {
+                this.value = value;
+                this.currentType = type;
+                return;
+            }
+        }
+        throw LOGGER.logThrowableAsError(new IllegalArgumentException("Invalid type: " + value.getClass().getName()));
+    }
+
+    /**
+     * Gets the type of the value.
+     *
+     * @return The type of the value.
+     */
+    public Type getType() {
+        return currentType;
+    }
+
+    /**
+     * Gets the types of the union.
+     *
+     * @return The types of the union.
+     */
+    public List<Type> getTypes() {
+        return Collections.unmodifiableList(types);
+    }
+
+    /**
+     * Gets the value of the union.
+     *
+     * @return The value of the union.
+     * @param <T> The type of the value.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getValue() {
+        return (T) value;
+    }
+
+    /**
+     * Gets the value of the union if it is of the expected type.
+     *
+     * @param clazz The expected type of the value.
+     * @return The value of the union.
+     * @param <T> The expected type of the value.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getValue(Class<T> clazz) {
+        if (clazz.isInstance(value)) {
+            return clazz.cast(value);
+        }
+        if (isPrimitiveTypeMatch(value, clazz)) {
+            return (T) value;
+        }
+        throw LOGGER.logThrowableAsError(new IllegalArgumentException("Value is not of type: " + clazz.getName()));
+    }
+
+    /**
+     * Gets the value of the union if it is of the expected type.
+     *
+     * @param clazz The expected type of the value.
+     * @param genericTypes The generic types of the expected type.
+     *
+     * @return The value of the union.
+     * @param <T> The expected type of the value.
+     */
+    public <T> T getValue(Class<T> clazz, Class<?>... genericTypes) {
+        return getValue(new ParameterizedTypeImpl(clazz, genericTypes));
+    }
+
+    /**
+     * Gets the value of the union if it is of the expected type.
+     *
+     * @param type The expected type of the value.
+     *
+     * @return The value of the union.
+     * @param <T> The expected type of the value.
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getValue(ParameterizedTypeImpl type) {
```

Comment:
Why do want to use the impl type instead of ParameterizedType as the param type?

Reply:
updated